File: 2015/java/7/Main.java
import java.io.File;
import java.util.Scanner;
import java.util.HashMap;
import java.util.ArrayDeque;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

enum InstructionType {
    AND,
    OR,
    LSHIFT,
    RSHIFT,
    NOT,
    ASSIGN
}

class Instruction {
    public InstructionType type;
    public String op1;
    public String op2;
    public String result;

    public Instruction(InstructionType type, String op1, String op2, String result) {
        this.type = type;
        this.op1 = op1;
        this.op2 = op2;
        this.result = result;
    }

    @Override
    public String toString() {
        return "Instruction [type=" + type + ", op1=" + op1 + ", op2=" + op2 + ", result=" + result + "]";
    }
}

public class Main {

    public static ArrayDeque<Instruction> instructions = new ArrayDeque<Instruction>();
    public static HashMap<String, Integer> registers = new HashMap<String, Integer>();

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java Main <input.txt>");
            System.exit(1);
        }

        String filename = args[0];
        File file = new File(filename);
        if (!file.exists()) {
            System.out.println("File not found: " + filename);
            System.exit(1);
        }

        try {

            Scanner scanner = new Scanner(file);
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                // use pattern matching to parse the line
                // and create an Instruction object
                Instruction instruction = null;
                String op1 = "";
                String op2 = "";
                String result = "";

                if (line.contains("AND") || line.contains("OR") || line.contains("LSHIFT") || line.contains("RSHIFT")) {
                    String regex = "(\\S+)\\s+(:?AND|OR|LSHIFT|RSHIFT)\\s+(\\S+)\\s+->\\s+(\\S+)";
                    Pattern pattern = Pattern.compile(regex);
                    Matcher matcher = pattern.matcher(line);

                    if (matcher.find()) {
                        op1 = matcher.group(1);
                        op2 = matcher.group(3);
                        result = matcher.group(4);
                    }

                    if (line.contains("AND")) {
                        instruction = new Instruction(InstructionType.AND, op1, op2, result);
                    }

                    if (line.contains("OR")) {
                        instruction = new Instruction(InstructionType.OR, op1, op2, result);
                    }

                    if (line.contains("LSHIFT")) {
                        instruction = new Instruction(InstructionType.LSHIFT, op1, op2, result);
                    }

                    if (line.contains("RSHIFT")) {
                        instruction = new Instruction(InstructionType.RSHIFT, op1, op2, result);

                    }
                    instructions.add(instruction);

                }

                if (line.contains("NOT")) {
                    String regex = "NOT\\s+(\\S+)\\s+->\\s+(\\S+)";
                    Pattern pattern = Pattern.compile(regex);
                    Matcher matcher = pattern.matcher(line);

                    if (matcher.find()) {
                        op1 = matcher.group(1);
                        result = matcher.group(2);
                    }

                    instruction = new Instruction(InstructionType.NOT, op1, "", result);
                    instructions.add(instruction);

                }

                if (instruction == null) {
                    String regex = "(\\S+)\\s+->\\s+(\\S+)";
                    Pattern pattern = Pattern.compile(regex);
                    Matcher matcher = pattern.matcher(line);

                    if (matcher.find()) {
                        op1 = matcher.group(1);
                        result = matcher.group(2);
                    }

                    instruction = new Instruction(InstructionType.ASSIGN, op1, "", result);
                    instructions.add(instruction);
                }

            }

            // Instruction a = instructions.stream().filter(i ->
            // i.result.equals("a")).findFirst().get();

            Integer a_result = eval("a");
            System.out.println("Part 1: " + a_result);

            registers.clear();
            registers.put("b", a_result);
            Integer a_result_part2 = eval("a");
            System.out.println("Part 2: " + a_result_part2);

            scanner.close();
        } catch (Exception e) {
            System.out.println(e);
            System.out.println("Error reading file: " + filename);
            System.exit(1);
        }
    }

    public static Integer eval(String target) {
        if (registers.containsKey(target)) {
            return registers.get(target);
        }

        Instruction instruction = null;

        for (Instruction i : instructions) {
            if (i.result.equals(target)) {
                instruction = i;
                break;
            }
        }

        Integer op1 = instruction.op1.matches("\\d+") ? Integer.parseInt(instruction.op1) : eval(instruction.op1);

        Integer op2 = 0;
        if (instruction.op2.length() > 0) {
            op2 = instruction.op2.matches("\\d+") ? Integer.parseInt(instruction.op2) : eval(instruction.op2);
        }

        Integer result = 0;
        switch (instruction.type) {
            case AND:
                result = op1 & op2;
                break;
            case OR:
                result = op1 | op2;
                break;
            case LSHIFT:
                result = op1 << op2;
                break;
            case RSHIFT:
                result = op1 >> op2;
                break;
            case NOT:
                result = ~op1;
                break;
            case ASSIGN:
                result = op1;
                break;
        }

        registers.put(target, result);
        return result;
    }
}

File: 2015/zig/7/src/main.zig
const std = @import("std");
const ArenaAllocator = std.heap.ArenaAllocator;
const assert = std.debug.assert;
const eql = std.mem.eql;
const string = @import("string");
const String = string.String;

const InstructionType = enum(u8) {
    AND = 0,
    OR,
    LSHIFT,
    RSHIFT,
    NOT,
    ASSIGN,
};

const Instruction = struct {
    type: InstructionType,
    lhs: []const u8,
    rhs: ?[]const u8,
    result: []const u8,
};

var instructions: std.ArrayList(Instruction) = undefined;
var hash_map: std.StringHashMap(u16) = undefined;

fn get_instruction(line: []const u8, line_buffer: *String) Instruction {
    var instr = Instruction{
        .type = InstructionType.ASSIGN,
        .lhs = undefined,
        .rhs = undefined,
        .result = undefined,
    };

    line_buffer.concat(line) catch {
        return instr;
    };

    if (line_buffer.find("AND") != null) {
        instr.type = InstructionType.AND;
        instr.lhs = line_buffer.split(" ", 0).?;
        instr.rhs = line_buffer.split(" ", 2).?;
        instr.result = line_buffer.split(" ", 4).?;
        return instr;
    }

    if (line_buffer.find("OR") != null) {
        instr.type = InstructionType.OR;
        instr.lhs = line_buffer.split(" ", 0).?;
        instr.rhs = line_buffer.split(" ", 2).?;
        instr.result = line_buffer.split(" ", 4).?;
        return instr;
    }

    if (line_buffer.find("LSHIFT") != null) {
        instr.type = InstructionType.LSHIFT;
        instr.lhs = line_buffer.split(" ", 0).?;
        instr.rhs = line_buffer.split(" ", 2).?;
        instr.result = line_buffer.split(" ", 4).?;
        return instr;
    }

    if (line_buffer.find("RSHIFT") != null) {
        instr.type = InstructionType.RSHIFT;
        instr.lhs = line_buffer.split(" ", 0).?;
        instr.rhs = line_buffer.split(" ", 2).?;
        instr.result = line_buffer.split(" ", 4).?;
        return instr;
    }

    if (line_buffer.find("NOT") != null) {
        instr.type = InstructionType.NOT;
        instr.lhs = line_buffer.split(" ", 1).?;
        instr.result = line_buffer.split(" ", 3).?;
        return instr;
    }

    if (line_buffer.find("->") != null) {
        instr.type = InstructionType.ASSIGN;
        instr.lhs = line_buffer.split(" ", 0).?;
        instr.result = line_buffer.split(" ", 2).?;
        return instr;
    }

    return instr;
}

fn get_target(target: []const u8) ?Instruction {
    for (instructions.items) |instr| {
        if (eql(u8, instr.result, target)) {
            return instr;
        }
    }

    return null;
}

fn evalOperand(operand: []const u8) u16 {
    // First, try to parse the operand as an integer literal
    const intResult = std.fmt.parseInt(u16, operand, 10);
    if (intResult) |value| {
        return value;
    } else |err| {
        switch (err) {
            std.fmt.ParseIntError.InvalidCharacter, std.fmt.ParseIntError.Overflow => {
                // If the operand is not an integer literal, try to evaluate it as a target
                return eval(operand);
            },
        }
    }
}

fn eval(target: []const u8) u16 {
    if (hash_map.contains(target)) {
        return hash_map.get(target).?;
    }

    const instr = get_target(target) orelse null;

    if (instr == null) {
        std.debug.print("Target: {s} not found\n", .{target});
        return 0;
    }

    var result: u16 = 0;

    const lhs = evalOperand(instr.?.lhs);
    var rhs: u16 = 0;
    if (instr.?.rhs) |rhs_operand| {
        rhs = evalOperand(rhs_operand);
    }

    switch (instr.?.type) {
        InstructionType.ASSIGN => {
            result = lhs;
        },
        InstructionType.AND => {
            result = lhs & rhs;
        },
        InstructionType.OR => {
            result = lhs | rhs;
        },
        InstructionType.LSHIFT => {
            result = lhs << @as(u4, @intCast(rhs));
        },
        InstructionType.RSHIFT => {
            result = lhs >> @as(u4, @intCast(rhs));
        },
        InstructionType.NOT => {
            result = ~lhs;
        },
    }

    hash_map.put(instr.?.result, result) catch {
        std.debug.print("Failed to insert key: {s} value: {d}\n", .{ instr.?.result, result });
    };
    return result;
}
pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    if (args.len < 2) {
        try stdout.print("Usage: zig run main.zig <input_file>\n", .{});
        return;
    }

    const file_path = args[1];
    const file = try std.fs.cwd().openFile(file_path, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    var in_stream = buf_reader.reader();

    var line_buf: [64]u8 = undefined;

    instructions = std.ArrayList(Instruction).init(std.heap.page_allocator);
    defer instructions.deinit();
    hash_map = std.StringHashMap(u16).init(std.heap.page_allocator);
    defer hash_map.deinit();

    var arena = ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();

    var buffers: std.ArrayList(String) = std.ArrayList(String).init(std.heap.page_allocator);
    defer buffers.deinit();

    while (true) {
        var line_buffer = String.init(arena.allocator());
        const line_bytes = try in_stream.readUntilDelimiterOrEof(&line_buf, '\n');
        if (line_bytes == null) {
            break;
        }

        if (line_bytes) |line| {
            // Remove the newline character
            const instr = get_instruction(line, &line_buffer);
            try instructions.append(instr);
        }

        try buffers.append(line_buffer);
    }

    var a = eval("a");
    try stdout.print("Part 1: {d}\n", .{a});
    hash_map.clearRetainingCapacity();
    hash_map.put("b", a) catch {
        std.debug.print("Failed to insert key: {s} value: {d}\n", .{ "b", a });
    };
    a = eval("a");
    try stdout.print("Part 2: {d}\n", .{a});

    for (0..buffers.items.len) |i| {
        buffers.items[i].deinit();
    }
}
